### PR TITLE
Camel instrumentation - SNS propagation impl + tests

### PR DESCRIPTION
**Pull request by kubawach**
- fixes issues with SNS to SQS propagation for Apache Camel instr
- added tests (currently ignored - work only with live AWS)

### CODE REVIEW

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+            "apache-camel.uri" "aws-sqs://${queueName}?amazonSQSClient=%23sqsClient"
+            "messaging.destination" queueName
+            "messaging.message_id" String
+          }
+        }
+      }
+      trace(7, 1) {
+        span(0) {
+          name "SQS.ReceiveMessage"
+          kind CLIENT
```
Comment (iNikem) [inline instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:268]: Why is it CLIENT here and CONSUMER above?

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+            "net.peer.name" String
+            "net.peer.port" {it == null || Number}
+            "net.transport" "IP.TCP"
+          }
+        }
+      }
+      trace(6, 3) {
+        span(0) {
+          name "SNS.Publish"
+          kind CLIENT
```
Comment (iNikem) [inline instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:218]: Should this be PRODUCER?

Comment (kubawach) [reply instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:218]: This is a tough question. 
In general, SNS message can be fan out to SQS (becoming a message), HTTP, lambda, kinesis or even email. In some of these cases we would have PUBLISH (send) - CLIENT (get) while now for SQS we have CLIENT (send) - CONSUMER (get). 
If we could, the best way IMO would be to modify "sending" span kind once a message is consumed (by SQS), but we can't. 
@anuraaga what do you think?

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy (hunk 1 above)
Comment (kubawach) [reply instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:268]: This is offspring of the modelling done for AWS SDK (as far as I remember). 
TLDR; each SQS response can have multiple messages, so each call (pooling) to SQS is modelled as a (HTTP) client, while each message consumed (if any) is a consumer span. Also, consumer spans are linked with producer spans as a single trace, while each SQS request is on its own.

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsConfig.groovy [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      @Override
+      void configure() throws Exception {
+        from("direct:input")
+          .log(LoggingLevel.INFO, "test", "SENDING body: \${body}")
+          .log(LoggingLevel.INFO, "test", "SENDING headers: \${headers}")
+          .to("aws-sns://snsCamelTest?amazonSNSClient=#snsClient")
+      }
+    }
+  }
+/**
```
Comment (mateuszrzeszutek) [inline instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsConfig.groovy:58]: Is this code needed?

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+            "aws.operation" "Publish"
+            "aws.service" "AmazonSNS"
+            "http.flavor" "1.1"
+            "http.method" "POST"
+            "http.status_code" 200
+            "http.url" String
+            "net.peer.name" String
+            "net.peer.port" {it == null || Number}
+            "net.transport" "IP.TCP"
+          }
```
Comment (mateuszrzeszutek) [inline instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:554]: Most of those AWS spans are almost the same, they only differ in names, kinds and `aws.operation` attributes - can you add a private method that encapsulates all other properties/attributes? This'll make those tests much shorter and easier to read

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy [hunk 5]
```diff
@@ -0,0 +1,696 @@
...
+    "    }]" +
+    "}"
+
+  def createAndSubscribeTopic(String topicName, String queueArn) {
+    CreateTopicResult ctr = snsClient.createTopic(topicName)
+    snsClient.subscribe(ctr.getTopicArn(), "sqs", queueArn)
+    return ctr.getTopicArn()
+  }
+
+  def "AWS SDK SNS producer - SQS consumer"() {
```
Comment (mateuszrzeszutek) [inline instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:71]:
```suggestion
  def "AWS SDK SNS producer - camel SQS consumer"() {
```

?

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy (hunk 4 above)
Comment (kubawach) [reply instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:554]: Yeah, I plan to clean-up all of Camel / AWS propagation tests once I'm done (I still want to test two additional scenarios). So another PR will include this 👍

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy (hunk 5 above)
Comment (kubawach) [reply instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:71]: +1

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by iNikem** [comment]
@kubawach Please rebase again. Main branch failure was fixed

File: instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy (hunk 2 above)
Comment (anuraaga) [reply instrumentation/apache-camel-2.20/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/apachecamel/SnsCamelTest.groovy:218]: From what I understand of the spec, we'd still generally expect the SNS message to be treated like messaging conventions so would be `PRODUCER`. The trickiness related to CLIENT and PRODUCER has come up before, I'm ok with just picking any for now but good chance of changing later.

**Review by anuraaga — APPROVED** [review]
